File: lib/queries/sync-settings.ts
import { api, unwrap } from './_client'

export interface SyncSettings {
  enabled: boolean
  centralUrl: string
  token: string
  restaurantId: string
  intervalSec: number
}

export async function fetchSyncSettings(): Promise<SyncSettings> {
  const r: any = await unwrap(api.GET('/api/v1/settings/sync'))
  return {
    enabled: !!r?.enabled,
    centralUrl: r?.central_url ?? '',
    token: r?.token ?? '',
    restaurantId: r?.restaurant_id ?? '',
    intervalSec: Number(r?.interval_sec ?? 30),
  }
}

export async function saveSyncSettings(s: SyncSettings): Promise<void> {
  await unwrap(api.PUT('/api/v1/settings/sync', {
    body: {
      enabled: s.enabled,
      central_url: s.centralUrl,
      token: s.token,
      restaurant_id: s.restaurantId,
      interval_sec: s.intervalSec,
    } as any,
  }))
}

// SyncQueueStats — состояние очереди отправки на central (ADR-003, Фаза О).
// Единственный способ для оператора убедиться, что накопленное за время без
// интернета действительно уехало.
export interface SyncQueueStats {
  pending: number
  failed: number
  oldestPendingAt?: string | null
  lastSyncedAt?: string | null
  lastError?: string | null
}

export async function fetchSyncQueueStats(): Promise<SyncQueueStats> {
  const r: any = await unwrap(api.GET('/api/v1/settings/sync/queue'))
  return {
    pending: Number(r?.pending ?? 0),
    failed: Number(r?.failed ?? 0),
    oldestPendingAt: r?.oldest_pending_at ?? null,
    lastSyncedAt: r?.last_synced_at ?? null,
    lastError: r?.last_error ?? null,
  }
}

// backfillSync — «Отправить историю на central» (ADR-003 Ф6): ставит в
// очередь ТЕКУЩЕЕ состояние каждой реплицируемой таблицы этого филиала,
// обычный пушер доставит на следующих циклах. Нужна для данных, заведённых
// ДО включения синхронизации (или мимо неё — напр. массовый импорт
// сотрудников, найдено 2026-08-29: central не видел официанта с 88 заказами,
// пока не отправили историю вручную). Идемпотентна, owner-only на бэке.
export interface SyncBackfillResult {
  entities: Record<string, number>
}

export async function backfillSync(): Promise<SyncBackfillResult> {
  const r: any = await unwrap(api.POST('/api/v1/sync/backfill', {}))
  return { entities: r?.entities ?? {} }
}

// joinNetwork — обменивает код приглашения (ADR-003, продолжение) на central
// на настоящий sync-токен+account_id и сохраняет всё атомарно на бэке;
// UI после успеха просто перечитывает fetchSyncSettings().
export async function joinNetwork(pairingCode: string): Promise<{ centralName: string }> {
  const r: any = await unwrap(api.POST('/api/v1/network/pair', { body: { pairing_code: pairingCode } as any }))
  return { centralName: r?.central_name ?? '' }
}

File: server/internal/service/sync_settings.go
package service

import (
	"context"
	"errors"
	"time"

	"gorm.io/gorm"
	"gorm.io/gorm/clause"

	"github.com/restos/restos-v4/server/internal/db/models"
	"github.com/restos/restos-v4/server/internal/repo"
)

// SyncSettingsService — конфиг multi-branch sync, редактируемый из UI (ADR-003/004).
// Не tenant-данные (конфиг узла), поэтому Raw + singleton (id=1).
type SyncSettingsService struct {
	r *repo.Repo
}

func NewSyncSettingsService(r *repo.Repo) *SyncSettingsService {
	return &SyncSettingsService{r: r}
}

// Get возвращает конфиг (или дефолты, если ещё не задан).
func (s *SyncSettingsService) Get(ctx context.Context) (*models.SyncSettings, error) {
	var st models.SyncSettings
	err := s.r.Raw().WithContext(ctx).Where("id = 1").First(&st).Error
	if errors.Is(err, gorm.ErrRecordNotFound) {
		return &models.SyncSettings{ID: 1, IntervalSec: 30}, nil
	}
	if err != nil {
		return nil, err
	}
	return &st, nil
}

// UpdateSyncSettingsInput — body PUT /api/v1/settings/sync.
type UpdateSyncSettingsInput struct {
	Enabled      bool   `json:"enabled"`
	CentralURL   string `json:"central_url"`
	Token        string `json:"token"`
	RestaurantID string `json:"restaurant_id"`
	IntervalSec  int    `json:"interval_sec"`
}

// Update сохраняет конфиг (upsert singleton). Enabled/CentralURL/Token/
// RestaurantID применяются в течение одного тика, БЕЗ перезапуска сервера —
// Pusher/Puller перечитывают sync_settings на каждом тике, не один раз при
// старте (ADR-003, продолжение; см. synclog.Pusher.activeConfig /
// service.Puller.activeConfig). Исключение — IntervalSec: частота самого
// тика фиксируется при старте процесса, смена применится со следующего
// перезапуска (сознательно не в объёме, редкая тонкая настройка).
func (s *SyncSettingsService) Update(ctx context.Context, in UpdateSyncSettingsInput) (*models.SyncSettings, error) {
	now := time.Now().UTC()
	interval := in.IntervalSec
	if interval <= 0 {
		interval = 30
	}
	st := &models.SyncSettings{
		ID:           1,
		Enabled:      in.Enabled,
		CentralURL:   strPtrOrNil(in.CentralURL),
		Token:        strPtrOrNil(in.Token),
		RestaurantID: strPtrOrNil(in.RestaurantID),
		IntervalSec:  interval,
		UpdatedAt:    &now,
	}
	// Явный список колонок, НЕ UpdateAll: backfilled_at (маркер «история уже
	// отправлена», Ф6) живёт в этой же строке, но им управляет только Backfill —
	// UpdateAll затирал бы его в NULL при каждом «Сохранить» из UI, и следующий
	// рестарт повторял бы полный забфилл впустую.
	if err := s.r.Raw().WithContext(ctx).
		Clauses(clause.OnConflict{
			Columns: []clause.Column{{Name: "id"}},
			DoUpdates: clause.AssignmentColumns([]string{
				"enabled", "central_url", "token", "restaurant_id", "interval_sec", "updated_at",
			}),
		}).
		Create(st).Error; err != nil {
		return nil, err
	}
	return st, nil
}

// SyncQueueStats — состояние очереди отправки (Фаза О). Нужна оператору:
// после дня без интернета единственный способ убедиться, что всё уехало, —
// это увидеть очередь. Раньше в UI не было ни одной цифры о синке, и
// «отправилось ли?» проверялось только по логам бэка.
type SyncQueueStats struct {
	// Pending — сколько дельт ждёт отправки прямо сейчас.
	Pending int64 `json:"pending"`
	// Failed — в карантине (central не принял, очередь пошла дальше без них).
	Failed int64 `json:"failed"`
	// OldestPendingAt — возраст головы очереди: если он растёт, синк стоит.
	OldestPendingAt *time.Time `json:"oldest_pending_at,omitempty"`
	// LastSyncedAt — когда в последний раз что-то реально уехало.
	LastSyncedAt *time.Time `json:"last_synced_at,omitempty"`
	// LastError — причина последней неудачи (из головы очереди), если есть.
	LastError *string `json:"last_error,omitempty"`
}

// QueueStats считает состояние очереди. Три дешёвых агрегата по sync_log:
// оба COUNT'а покрыты частичным индексом idx_sync_log_pending.
func (s *SyncSettingsService) QueueStats(ctx context.Context) (*SyncQueueStats, error) {
	out := &SyncQueueStats{}
	db := s.r.Raw().WithContext(ctx)

	if err := db.Model(&models.SyncLog{}).
		Where("synced_at IS NULL AND failed_at IS NULL").Count(&out.Pending).Error; err != nil {
		return nil, err
	}
	if err := db.Model(&models.SyncLog{}).
		Where("failed_at IS NOT NULL").Count(&out.Failed).Error; err != nil {
		return nil, err
	}

	var head models.SyncLog
	if err := db.Where("synced_at IS NULL AND failed_at IS NULL").
		Order("created_at ASC").First(&head).Error; err == nil {
		out.OldestPendingAt = &head.CreatedAt
		out.LastError = head.LastError
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}

	var last models.SyncLog
	if err := db.Where("synced_at IS NOT NULL").
		Order("synced_at DESC").First(&last).Error; err == nil {
		out.LastSyncedAt = last.SyncedAt
	} else if !errors.Is(err, gorm.ErrRecordNotFound) {
		return nil, err
	}
	return out, nil
}

func strPtrOrNil(s string) *string {
	if s == "" {
		return nil
	}
	return &s
}
